File: java/strings/medium/LongestRepeatingCharacterReplacement.java
class Solution {
    public int characterReplacement(String s, int k) {
        int n = s.length();
        int left = 0;
        int max_length = 0;
        int max_count = 0;
        int[]chars = new int[26];
        
        for (int right = 0; right < n; right++){
            chars[s.charAt(right)-'A']++;
            int current_char_count = chars[s.charAt(right)-'A'];
            max_count = Math.max(max_count,current_char_count);
            
            while (right - left - max_count + 1 > k) {
                chars[s.charAt(left)-'A']--;
                left++;
            }
            
            max_length = Math.max(max_length, right - left + 1);
        }
        
        return max_length;
        
    }
}

File: Python/Strings/Medium/longestRepeatingCharacterReplacement.py
class Solution(object):
    def longestRepeatingCharacterReplacement(self, s, k):
        # map to count the occurrences of characters
        count = {}
        length = len(s)

        # init left pointer, max_length, max_frequency
        left, max_length, max_frequency = 0, 0, 0

        # loop through s 
        for right in range(length):

            # get count of chars
            count[s[right]] = 1 + count.get(s[right], 0)
            max_frequency = max(max_frequency, count[s[right]])

            # check if window is valid, decrement count of char, and move left pointer 
            while (right - left + 1) - max_frequency > k:
                count[s[left]] -= 1
                left += 1

            # update max_length 
            max_length = max(max_length, right - left + 1)

        return max_length 

# Time comp: O(n)
# Space comp: O(n)